ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -683,7 +716,154 @@ public void handleReplicaMetadataRequest(NetworkRequest request) throws IOExcept
+              getResponse.getInputStream(), orgMsgInfo);
+          if (output == null || !output.getMessageInfo().getStoreKey().equals(blobId)) {
+            logger.error("ReplicateBlobRequest transferInputStream {} returned null, {} {} {}",
+                orgMsgInfo, remoteHostName, remoteHostPort, blobId);
+            throw new StoreException("ReplicateBlobRequest " + replicateBlobRequest + " transferInputStream returned "
+                + output, StoreErrorCodes.Unknown_Error);
+          } else {
+            MessageFormatWriteSet writeset = new MessageFormatWriteSet(output.getStream(),
+                                            Collections.singletonList(output.getMessageInfo()), false);
+            Store store = storeManager.getStore(blobId.getPartition());
+            store.put(writeset);
+
+            // As ReplicateThread, we do applyTtlUpdate and also applyDelete.
+            if (orgMsgInfo.isTtlUpdated()) {
+              applyTtlUpdate(orgMsgInfo, replicateBlobRequest);
+            }
+            if (orgMsgInfo.isDeleted()) {
+              applyDelete(orgMsgInfo, replicateBlobRequest);
+            }
+            logger.info("ReplicateBlobRequest replicated Blob {} from remote host {} {}", blobId, remoteHostName, remoteHostPort);
+            response = new ReplicateBlobResponse(replicateBlobRequest.getCorrelationId(), replicateBlobRequest.getClientId(),
+                ServerErrorCode.No_Error);
+          } // if (output == null) {
+        } // if (messageInfoList == null || messageInfoList.size() != 1)
+      } catch (StoreException e) {
+        if (e.getErrorCode() == StoreErrorCodes.Already_Exist) {
+          logger.info("ReplicateBlobRequest Blob {} already exists for {}", blobId, replicateBlobRequest);
+          response = new ReplicateBlobResponse(replicateBlobRequest.getCorrelationId(), replicateBlobRequest.getClientId(),
+              ServerErrorCode.No_Error);
+        } else {
+          logger.error("ReplicateBlobRequest unknown exception to replicate {} of {}", blobId, replicateBlobRequest, e);
+          response = new ReplicateBlobResponse(replicateBlobRequest.getCorrelationId(), replicateBlobRequest.getClientId(),
+              ServerErrorCode.Unknown_Error);
+        }
+      }
+    } else if (partitionResponseInfo.getErrorCode() == ServerErrorCode.Blob_Deleted) {
+      // Since GetOption is Include_All, even it's deleted on the remote replica, we'll still get the PutBlob.
+      // One exception is that because of compaction or other reasons, the PutRecord is gone.
+      logger.error("ReplicateBlobRequest Blob {} of {} is delete from the source replica", blobId, replicateBlobRequest);
+      response = new ReplicateBlobResponse(replicateBlobRequest.getCorrelationId(), replicateBlobRequest.getClientId(),
+          ServerErrorCode.Blob_Not_Found);
+    } else if (partitionResponseInfo.getErrorCode() == ServerErrorCode.Blob_Authorization_Failure) {
```

Comment:
QQ: Would we need to handle this error case. Since it is server-server communication, would we get authorization failure?

Reply:
I'm not sure about that. Probably you are right, we won't hit this case.
But since replication code handles it, I keep it here for safe.
I hope it's ok.

Reply:
Removed it.